packet_filter.py

```diff
@@ -64,15 +64,21 @@ def detect_tls_events(input_stream):
 
 
 def warn_about_ip_address(ip_address, bad_ips):
-    if ip_address in bad_ips:
-        ioc_source = bad_ips[ip_address]
-        logging.info('CONNECTING WITH BAD IP %s (found in %s)',
-                     ip_address, ioc_source)
+  ip_address = str(ip_address)
+  if ip_address in bad_ips:
+    ioc_sources = bad_ips[ip_address]
+    logging.info('CONNECTING WITH BAD IP %s (found in %s)',
+                 ip_address, ioc_sources)
+    return True
+  return False
  
 
-def detect_bad_ips(input_stream, bad_ips):
-    for line in iter(input_stream, b''):
-        values = line.split('\t')
-        ip_from, ip_to = ipaddress.ip_address(values[1]), ipaddress.ip_address(values[2])
-        check_ip_address(ip_from, bad_ips)
-        check_ip_address(ip_to, bad_ips)
+def detect_bad_ips(dataline, bad_ips):
+  if not dataline:
+    return False
+  values = dataline.split('\t')
+  ip_from, ip_to = ipaddress.ip_address(values[1]), ipaddress.ip_address(values[2])
+  if (warn_about_ip_address(ip_from, bad_ips)
```

Comment:
Why are we differentiating from and to ips? Aren't all ips "to"?

I thought we where going to do connection-based detection, not packet based

Reply:
It depends on the direction of the packet, the naming is taken directly from tshark's output and 'from' and 'to' are in flipped directions when the packet is going back to the client.  And, in this case, we should be just as concerned if a client is using an IP address that has been marked as compromised or hostile in a public IOC list, I think.

We were using connection-based detection for the suspicious TLS hellos but for the known IOCs being integrated we can benefit from including connectionless traffic, too.  I'm sure there is malware that uses UDP that we would want to be aware of.

Reply:
I think we could still capture packets from the network device even if integrating with Outline?  If we want it to also be an IPS then yes, we might have to constrain ourselves to the network flow view that Outline has, but I thought we already had this design discussion, and preferred this approach as well as not making design decisions based on how it might integrate with Outline.

Our packet-focused approach is largely a result of using tshark to do the network traffic capture.  We could still stitch streams together but a lot of IOCs favor a packet-based approach too.  Even the TLS hello inspection we do is packet-based, except perhaps the part where we connect the client hello and server hello data together.  Even that part I have some misgivings about because there are client hellos that don't get a response but might still be interesting (e.g. a malware that is trying to connect to a C2 that has since been dismantled).

I'll add some clarification as pydoc on these functions.